### ISSUE TITLE
Editor Height Smoothing Tool Occasionally Causes Unfixable Craters 

### ISSUE THREAD

**Issue by Romelete** [SmartlyDressedGames/Unturned-3.x-Community#4006]
When using the smooth tool, it will occasionally suddenly dip the terrain to zero.  This cannot be fixed using any other tool.
![image](https://github.com/SmartlyDressedGames/Unturned-3.x-Community/assets/33403276/6df844f1-7706-4240-a30c-5ef09198abb4)
clicking anywhere along the edges results in the crater spreading very quickly, sometimes crossing the entire map in seconds, resetting all terrain in its path to zero.
![image](https://github.com/SmartlyDressedGames/Unturned-3.x-Community/assets/33403276/e0858da2-6ca9-4f69-87c4-3e2d182b44d9)
![image](https://github.com/SmartlyDressedGames/Unturned-3.x-Community/assets/33403276/c238f437-219c-4d4f-931f-079cae319dfb)

edit: tool settings
![image](https://github.com/SmartlyDressedGames/Unturned-3.x-Community/assets/33403276/527ee87e-a382-406e-938f-a6aeab1aa836)

edit 2: exiting the map editor and re-entering allows you to fix it with other tools, but it's still very frustrating

**Comment by SDGNelson** [comment]
Sorry about this, it's proven quite a mystery. I was able to reproduce it **one** time but never again, and that was without any extra logging. Have you noticed any pattern in being able to recreate it?

**Comment by Romelete** [comment]
the last few times it's happened was using the smooth tool with a very large brush size, but it seems to be very sporadic as to when it actually happens so it's hard to tell

**Comment by DanielWillett** [comment]
I will say I've been working with the terrain tools a lot lately and haven't noticed this (under normal circumstances), however when the heights were being set to NaN or any other non-real numbers, it would cause this effect. Maybe check for places where division by zero is possible?

**Comment by AdamDN** [comment]
> Sorry about this, it's proven quite a mystery. I was able to reproduce it **one** time but never again, and that was without any extra logging. Have you noticed any pattern in being able to recreate it?

It can be very consistently reproduced when working with extreme high/lows and then smoothing with a big brush.

https://github.com/SmartlyDressedGames/Unturned-3.x-Community/assets/71527679/82c06144-c3e9-459f-8dfd-0cf6331db9a4

But it also happens (less frequently) in pretty reasonable scenarios like this, again with a large smooth tool brush.

https://github.com/SmartlyDressedGames/Unturned-3.x-Community/assets/71527679/18fe7e41-d152-4e36-926c-325e504eefc9


Like OP said, the effected terrain can't be fixed with any tools until you save, exit and reload the map. (or undo)

**Comment by SDGNelson** [comment]
Thanks guys, when I checked into it before I thought it might be division by zero but didn't see how it could happen for example in the heightmap smoothing function. AdamDN I was able to recreate it using your example (thanks!) and found it was NaN, but not for the expected reason:

`currentHeight: 0.5131553 heightmapSmoothTarget: 0.5131466 heightmapSmoothSampleAverage: 2.052587 heightmapSmoothSampleCount: 4 deltaTime: 0.0189329 strength: 16 alpha: NaN`

Brush alpha NaN? 😕 I hadn't looked there for this bug...
```
if (normalizedDistance < brushFalloff)
{
	return 1.0f;
}
else
{
	return (1.0f - normalizedDistance) / (1.0f - brushFalloff);
}
```
If the point happened to be right at the edge of the brush with 1.0 brush falloff it would divide by zero! Hopefully that's the only case that was causing this. Fortunately in your video it looks like you were using falloff 1.0.
